Tutorials/13_Geometry-Optimization/opt_helper/oofp.py

```diff
@@ -0,0 +1,135 @@
+from math import cos, sin, tan
+
+from . import optExceptions
+from . import v3d
+from .simple import *
+
+FIX_VAL_NEAR_PI = 1.57
+from psi4 import constants
+HARTREE2AJ = constants.hartree2aJ
+
+# Class for out-of-plane angle.  Definition (A,B,C,D) means angle AB with respect
+# to the CBD plane; canonical order is C < D
+
+
+class OOFP(SIMPLE):
+    def __init__(self, a, b, c, d, frozen=False, fixedEqVal=None):
+
+        if c < d: atoms = (a, b, c, d)
+        else: atoms = (a, b, d, c)
+        self._near180 = 0
+        SIMPLE.__init__(self, atoms, frozen, fixedEqVal)
+
+    def __str__(self):
+        if self.frozen: s = '*'
+        else: s = ' '
+
+        s += "O"
+
+        s += "(%d,%d,%d,%d)" % (self.A + 1, self.B + 1, self.C + 1, self.D + 1)
+        if self.fixedEqVal:
+            s += "[%.4f]" % self.fixedEqVal
+        return s
+
+    def __eq__(self, other):
+        if self.atoms != other.atoms: return False
+        elif not isinstance(other, OOFP): return False
+        else: return True
+
+    @property
+    def near180(self):
+        return self._near180
+
+    def updateOrientation(self, geom):
+        tval = self.q(geom)
+        if tval > FIX_VAL_NEAR_PI:
+            self._near180 = +1
+        elif tval < -1 * FIX_VAL_NEAR_PI:
+            self._near180 = -1
+        else:
+            self._near180 = 0
+        return
+
+    @property
+    def qShowFactor(self):
+        return 180.0 / np.pi
+
+    def qShow(self, geom):  # return in degrees
+        return self.q(geom) * self.qShowFactor
+
+    @property
+    def fShowFactor(self):
+        return HARTREE2AJ * np.pi / 180.0
+
+    # compute angle and return value in radians
+    def q(self, geom):
+        check, tau = v3d.oofp(geom[self.A], geom[self.B], geom[self.C], geom[self.D])
+        if not check:
+            raise optExceptions.ALG_FAIL(
+                "OOFP::compute.q: unable to compute out-of-plane value")
```

Comment:
It would be more usual for the fns in v3d that hit some tolerance that currently results in a `check = False` to instead raise an exception right when the error occurs in v3d, rather than passing the flag back up to here to handle. That way the calling function could proceed assuredly. Is there a reason to defer error handling back to the calling fn, like the above?

Reply:
Got it. I'm far from demanding it be changed. The pattern I notice is that right now, some of the fns are returning either `False, some_innocent_default_value` or `True, computed value`. That is, in either case, only one piece of useful info is actually returned – `False` for the former, `computed_value` for the latter. I do suspect that exceptions over flags would be clearer; does the below seem a helpful pattern?

```
def oofp():
    first step
    if something_wrong:
        raise MajorProblemError(msg)
    next_step
    if something_wrong:
        raise SlightlyIllConditionedError(msg)
    final step
    return ans

class OOFP(object):
    def doing_something:
        try:
            ans = oofp()
        except SlightlyIllConditionedError as err:
            ans = massage_a_bit()
        # no need to handle MajorProblemError as execution ended
        next_step(ans)
```